### ISSUE TITLE
A route named 'apiv2package_upload' is already in the route collection

### ISSUE THREAD

**Issue by Levente0xFFFF** [NuGet/NuGetGallery#7008]
[This change breaks our application.](https://github.com/NuGet/NuGet.Server/pull/62)

We cannot upgrade from NuGet.Server.V2 3.1.2 to 3.2.0 because of this change. This is a breaking change, so this package should be unlisted and re-released with version 4.0.0. See semver.org for details.

We are getting this exception when calling NuGetV2WebApiEnabler.UseNuGetV2WebApiFeed():

> System.ArgumentException: 'A route named 'apiv2package_upload' is already in the route collection. Route names must be unique.
> Parameter name: name'

We are hosting multiple NuGet feeds on the same web app and passing the routeName as a parameter.
Since this change hard codes the route name this is not possible anymore:
https://github.com/NuGet/NuGet.Server/blob/e5d8ee21632a1ce1d7a226e0b1a1060277249cf6/src/NuGet.Server.V2/NuGetV2WebApiEnabler.cs#L57

All the above route registrations are taking the routeName parameter into consideration, but not this one.

**Comment by joelverhagen** [comment]
Copied from https://github.com/NuGet/NuGet.Server/pull/62#issuecomment-474825132.

> @Levente0xFFFF, thanks for the report and we apologise for the inconvenience. Could you open a new GitHub issue so we can track this separately?
>
> It seems like a straightforward fix would be to make this route optional. Do you think a bool parameter to this method would be sufficient to solve your problem?

**Comment by Levente0xFFFF** [comment]
Adding a new parameter to an existing method will be a breaking change in API compatibility, the original method without the new bool parameter will be deleted even if the new bool parameter is optional. Since I will republish my app anyways, it is a not a problem for me, but I don't know about other people, how they use this DLL, because a drop in replacement (copy and overwrite upgrade) won't be possible anymore. Recompiling and republishing the project will be needed, this can be a problem if the source is lost or if they don't want to do it for any reason.

Adding a new method could work without any drawbacks though, it could be called: UseNuGetV2WebApiFeedTheEricColemanStyle.

Maybe ericcoleman should just fork this repo and use his own DLL if nobody else is affected by his issue.

**Comment by joelverhagen** [comment]
Totally understand the APi compatibility issue. Given the breaking change already shipped we would be choosing between breaking one default vs. another. I personally am leaning towards the latter approach you mentioned (require a bool parameter to activate this route) since this more aligns with rest of the 3.x releases. 

I think forking is not necessary here since the backwards compatible route helps old 2.x customers and in my opinion should be available _somehow_ in 3.x.

We'll consider the options and respond back on this issue when we have made some progress.

**Comment by joelverhagen** [comment]
Oh and to be clear I meant a required bool parameter on a new method. Old method with no bool would remain with a behavior we choose (probably the pre-3.2.0 behavior).

**Comment by Levente0xFFFF** [comment]
There is another option: unlist. Since this package is only out since 1 week and the impact of this bug is very severe, this could be considered also.
Unlist 3.2.0, then release 3.3.0 (or 3.2.1) with the following:
Create a new private method which does the actual thing, it would have the new boolean parameter.
The original UseNuGetV2WebApiFeed method would just call the new private method with the boolean parameter set to false.
The new public UseNuGetV2WebApiFeed method with the new boolean parameter would call the private method with the value of the boolean parameter, which in practice would be true in almost every case, because otherwise the user could just call the old original method.

**Comment by joelverhagen** [comment]
Yep that sounds like a good plan. I was thinking something along those lines. I also think unlisting the 3.2.0 version is a good move. A new minor version is not really necessary in this case. We can just incremental patch version to 1. The semantics of a patch version in conjunction with an unlisted version it's kind of undefined but I think it could fit here.

**Comment by joelverhagen** [comment]
Hey @Levente0xFFFF, could you try out the following package to see if it solves your problem?

Package source: https://dotnet.myget.org/F/nuget-build/api/v3/index.json
Package version: 3.2.1-jver-7008-2513641

You can get NuGet.Server.Core, NuGet.Server.V2, and NuGet.Server from here.

**Comment by joelverhagen** [comment]
Note that this is just a pre-release build to validate that your problem is resolved with the proposed fix. After you have confirmed, I'll move forward with getting this package on nuget.org as version 3.2.1 and unlisting 3.2.0.

**Comment by Levente0xFFFF** [comment]
I can confirm that it is working as expected.
I also checked what happens if I use the new overload with enableLegacyPushRoute: true. I get the expected exception.

The code change also looks perfect, I see that the private method I mentioned is not even needed, I don't know why I wrote that it is needed.

Thank you for the prompt response and fix.

Although, if @ericcoleman intended his fix as a DLL drop in replacement, it won't work for him anymore, he needs to modify and recompile his app, but that's fine. He can still fork if he doesn't want to :)
